dump/main.py

```diff
@@ -0,0 +1,214 @@
+
+# Autocompletion Helper
+def get_available_modules(ctx, args, incomplete):
+    return [k for k in plugins.dump_modules.keys() if incomplete in k]
+
+# Display Modules Callback
+def show_modules(ctx, param, value):
+    if not value or ctx.resilient_parsing:
+        return
+    header = ["Module", "Identifier"]
+    display = []
+    for mod in plugins.dump_modules:
+        display.append((mod, plugins.dump_modules[mod].ARG_NAME))
+    click.echo(tabulate(display, header))
+    ctx.exit()
+             
+@click.group()
+def dump():
+    pass
+
+@dump.command()
+@click.pass_context
+@click.argument('module', required=True, type=str, autocompletion=get_available_modules)
+@click.argument('identifier', required=True, type=str) 
+@click.option('--show', '-s', is_flag=True, default=False, help='Display Modules Available', is_eager=True, expose_value=False, callback=show_modules)
+@click.option('--db', '-d', multiple=True, help='Only dump from these Databases or the CONFIG_FILE')
+@click.option('--table', '-t', is_flag=True, default=False, help='Print in tabular format', show_default=True)
+@click.option('--key-map', '-k', is_flag=True, default=False, help="Only fetch the keys matched, don't extract field-value dumps", show_default=True)
+@click.option('--verbose', '-v', is_flag=True, default=False, help="Prints any intermediate output to stdout useful for dev & troubleshooting", show_default=True)
+@click.option('--namespace', '-n', default=DEFAULT_NAMESPACE, type=str, show_default=True, help='Dump the redis-state for this namespace.')  
+def state(ctx, module, identifier, db, table, key_map, verbose, namespace):
+    """
+    Dump the current state of the identifier for the specified module from Redis DB or CONFIG_FILE
+    """
+    if not multi_asic.is_multi_asic() and namespace != DEFAULT_NAMESPACE:
+        click.echo("Namespace option is not valid for a single-ASIC device")
+        ctx.exit()
+    
+    if multi_asic.is_multi_asic() and (namespace != DEFAULT_NAMESPACE and namespace not in multi_asic.get_namespace_list()):
+        click.echo("Namespace option is not valid. Choose one of {}".format(multi_asic.get_namespace_list()))
+        ctx.exit()
+    
+    if module not in plugins.dump_modules:
+        click.echo("No Matching Plugin has been Implemented")
+        ctx.exit()
+    
+    if verbose:
+        os.environ["VERBOSE"] = "1"
+    else:
+        os.environ["VERBOSE"] = "0"
+    
+    ctx.module = module
+    obj = plugins.dump_modules[module]()
+    
+    if identifier == "all":
+        ids = obj.get_all_args(namespace)
+    else:
+        ids = identifier.split(",")
+        
+    params = {}
+    collected_info = {}
+    params['namespace'] = namespace
+    for arg in ids: 
+        params[plugins.dump_modules[module].ARG_NAME] = arg 
+        collected_info[arg] = obj.execute(params)
+        
+    if len(db) > 0:
+        collected_info = filter_out_dbs(db, collected_info)
```

Comment:
Instead of filtering out the db's here. Is it possible to collect info only the DBs the user has specified. 
This may reduce some processing overhead ?

Reply:
No, to filter all the keys as expected, for eg: keys in ASIC, we might need information to be fetched from CONFIG or APPL DB. So, the traversal across db's is not easily decoupled. Hence decided to filter it once all the required keys are fetched